Add the ability to forward a base64 encoded cert in the xfcc header.

jhspaybar opened the pull request:
Signed-off-by: William Thurston <wthurston@netflix.com>

*tls*: *add the ability to forward a base64 encoded cert in the xfcc header.*


*Description*:
This feature allows the forwarding of the entire cert base64 encoded when mTLS is present and APPEND_FORWARD is enabled in tls settings.  This allows the application layer to inspect things like extensions, or OID fields which may be needed for additional AuthZ decisions.

*Risk Level*: Low

*Testing*:
Added unit tests which assert presence and absence based on configuration changes.  Manually tested this with the option forced to true.  The cert is present in the header as expected and could be parsed in Java.

*Docs Changes*:
https://github.com/envoyproxy/data-plane-api/pull/390

*Release Notes*:
An additional option is present which allows for the forwarding of an entire cert base64 encoded in the x-forwarded-client-cert header.  This can allow for application level inspection of the cert if needed.

mattklein123 commented:
@jhspaybar quick note that you can refer to data-plane-api in your fork temporarily if you want to work on tests, etc.

jhspaybar commented:
Thanks, I'll try referencing my local branch and see if I can get some tests going before that other PR is merged.

jhspaybar commented:
Also, this is my first real c++ code ever, first time using openssl, committing to envoy, etc.  I've done my best not to leak memory or cause other problems, and if there are better or more idiomatic ways to do things feel free to ask me to make those changes, I'm learning how to write production c++ as I go.

asibross commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -226,6 +227,36 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  }
+
+  int length;
+  unsigned char *str = NULL;
```
Minor: use nullptr instead of NULL.

asibross commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -226,6 +227,36 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+    return Envoy::Base64::encode(mem->data, mem->length);
+  }
+  
+  return "";
```
Consider rearranging the mem checks to return "" and return the actual string as the last statement:

Something like this:
BUF_MEM *mem = nullptr;
BIO_get_mem_ptr(buf.get(), &mem);
if (!mem || !mem->data || (mem->length == 0)) {
  return "";
}

return Envoy::Base64::encode(mem->data, mem->length);
}

asibross commented on source/server/config/network/http_connection_manager.cc:
```diff
@@ -158,6 +158,9 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
   }
+  /*if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(set_current_client_cert_details, cert, false)) {
+    set_current_client_cert_details_.push_back(Http::ClientCertDetailsType::Cert);
+  }*/
```
Uncomment or remove.

jhspaybar commented:
Updated to address comments from @asibross .  I'll work on referencing the new config properly so I can get some tests added.

jhspaybar commented:
Okay, updated again to use the new proto field.  I also tweaked some existing tests to have assertions with and without the base64 encoded cert.  How to modify or extend the xfcc integration tests weren't completely clear to me, but I could revisit and take a shot at those if it makes sense to do so.  I did also manually verify that I can decode the base64 cert in Java land just to be safe.

```if (headers.get(X509_KEY) != null) {
                        String[] split = headers.get(X509_KEY).split(";");
                        String b64 = null;
                        for (String item : split) {
                            if(item.startsWith("Cert=\"")) {
                                b64 = item.substring(6, item.length() - 1);
                            }
                        }
                        if (b64 != null) {
                            byte encodedCert[] = Base64.getDecoder().decode(b64);
                            ByteArrayInputStream inputStream = new ByteArrayInputStream(encodedCert);
                            CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
                            X509Certificate cert = (X509Certificate) certFactory.generateCertificate(inputStream);
                            System.out.println("Cert is: " + cert.toString());
                        }
                    }```
This does lead me to one question though, it isn't clear to me what is a safe and proper way to actually parse this header.  Splitting on ";" is probably safe, but if there's a ";" in the subject for example, I could end up in a weird spot, Especially if it's something like "My Subject Name;Cert=...."  Should we potentially be encoding this into the header as a base64 of a protobuf object so there is no ambiguity?

dnoe commented on include/envoy/ssl/connection.h:
```diff
@@ -42,6 +42,12 @@ class Connection {
   virtual std::string uriSanPeerCertificate() PURE;
+
+  /**
+   * @return the base64 encoded representation of the peer certificate. Returns "" if there is no
```
Envoy style nit, return type goes in return comment: `@return std::string the base64 encoded...`

dnoe commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -188,11 +188,13 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
       config.forwardClientCert() == Http::ForwardClientCertType::SanitizeSet) {
-    if (!connection.ssl()->uriSanLocalCertificate().empty()) {
-      client_cert_details.push_back("By=" + connection.ssl()->uriSanLocalCertificate());
+    std::string uri_san_local_cert = connection.ssl()->uriSanLocalCertificate();
```
`const std::string`

dnoe commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -188,11 +188,13 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
     }
-    if (!connection.ssl()->sha256PeerCertificateDigest().empty()) {
-      client_cert_details.push_back("Hash=" + connection.ssl()->sha256PeerCertificateDigest());
+    std::string cert_digest = connection.ssl()->sha256PeerCertificateDigest();
```
`const std::string`

dnoe commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -205,6 +207,13 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
         client_cert_details.push_back("SAN=" + connection.ssl()->uriSanPeerCertificate());
+        break;
+      case Http::ClientCertDetailsType::Cert:
+        std::string peer_cert = connection.ssl()->peerCertificate();
```
Use a const string here too.

jhspaybar commented:
Thanks @dnoe addressed your comments.  I'll wait for an SSL expert to weigh in :)

dnoe commented:
Thanks! And FYI the default preference for Envoy PRs is to add new commits rather than rebase/amend/force push because it makes the subsequent reviews easier in the GitHub UI.

jhspaybar commented:
Good to know, do you like me to squash at the end so there's just a single commit?

dnoe commented:
Whichever Envoy maintainer does the final merge will do the squash through GitHub when the PR is merged. Make sure that the DCO sign-off (signed-off-by) is present in each commit to avoid annoying the DCO bot. Thanks!

mattklein123 commented on include/envoy/ssl/connection.h:
```diff
@@ -15,33 +15,39 @@ class Connection {
-   * @return the URI in the SAN feld of the local certificate. Returns "" if there is no local
-   *         certificate, or no SAN field, or no URI.
+   * @return std::string the URI in the SAN feld of the local certificate. Returns "" if there is no
+   *local certificate, or no SAN field, or no URI.
```
super nit: missing space after '*'. Same in several other places below.

jhspaybar replied:
So I added std::string from the style nit above to these doc strings, and it made the line too long.  So I ran the style fix script, and that is what formatted these comments like this.  I can go manually add the extra space(I agree, it's more aesthetic), but we may want to get it in the fix script, or at least not have the fix script do this when the column width is too large on the previous line.

mattklein123 commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -226,6 +227,35 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  int length;
+  unsigned char* str = nullptr;
+  length = ASN1_item_i2d(reinterpret_cast<ASN1_VALUE*>(cert.get()), &str, ASN1_ITEM_rptr(X509));
+  if (!str) {
```
I will defer to @ggreenway or @PiotrSikora on this boring ssl code here, but in general we prefer to use `RELEASE_ASSERT` and crash for calls that should never reasonably fail (please see the error handling guide in STYLE.md). I think everything other than the initial call to get the peer certificate is in that category.

mattklein123 commented:
> This does lead me to one question though, it isn't clear to me what is a safe and proper way to actually parse this header. Splitting on ";" is probably safe, but if there's a ";" in the subject for example, I could end up in a weird spot, Especially if it's something like "My Subject Name;Cert=...." Should we potentially be encoding this into the header as a base64 of a protobuf object so there is no ambiguity?

Hmm. @myidpt @lizan ^^^ WDYT? What are you doing in Istio?

ggreenway commented:
I agree with @PiotrSikora that this is a weird format.  Just send the PEM format, which is already base64.

ggreenway commented:
Also, unless PEM format will be compatible with the nginx variable $ssl_client_cert.  And if we want to learn from issues nginx has run into, you can URL-encode it like $ssl_client_escaped_cert.

jhspaybar commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -188,11 +188,13 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
       config.forwardClientCert() == Http::ForwardClientCertType::SanitizeSet) {
-    if (!connection.ssl()->uriSanLocalCertificate().empty()) {
-      client_cert_details.push_back("By=" + connection.ssl()->uriSanLocalCertificate());
+    const std::string uri_san_local_cert = connection.ssl()->uriSanLocalCertificate();
```
I did no exhaustive testing of performance.  I only noticed that we were creating these strings twice and could create them once instead.  I noticed when I wrote my code in the format of a single call and then realized it looked different.  There were also some tests that asserted it was called twice, but I don't think it was needed for correctness or anything.  I could change it back, but it looks easy/opportunistic and let my code no longer stand out as different.

myidpt replied:
I see. I was just curious. I think from the test point of view calling it only once seems better. You can keep it in the new way.

jhspaybar commented:
@PiotrSikora @ggreenway I have no strong feelings about how the cert is passed.  I started with the PEM format early on and realized that there would be newlines and such in the header.  I suspect this is why nginx now URL-encodes it is to get a single clean line for the value.  I went the route of just base64 encoding bytes because it seemed like the most direct path to a single line and easily parsed universal format in a header.  If you'd like it to be a URL-encoded PEM instead, I'm also happy to do that.

jhspaybar commented:
Before I make changes to the encoding, let me know the direction we'd like to go.  I went with this format because I didn't have to deal with newlines in a header, I could go the PEM approach and either URL encode it like nginx, or replace newlines with spaces(this seems to be what at least one way of forwarding the cert that apache can do).

mattklein123 commented:
@jhspaybar I don't have a strong opinion, but I would probably default to just doing what nginx does unless there is a good reason not to. @ggreenway @PiotrSikora ?

ggreenway commented:
> @jhspaybar I don't have a strong opinion, but I would probably default to just doing what nginx does unless there is a good reason not to. @ggreenway @PiotrSikora ?

That's my feeling also.

PiotrSikora commented:
Quite frankly, I forgot about new lines when suggesting PEM, sorry!

I don't have any preference between `base64(DER)` and `urlencode(PEM)`, so either one is fine with me (which means switching to `urlencode(PEM)`, per @ggreenway and @mattklein123).

ggreenway commented:
This will need a master merge to resolve conflicts.

jhspaybar commented:
I asked in Slack, but hadn't received an answer yet, and I was wondering if there already exists a method somewhere in the envoy codebase that would URL encode a string?  I'd prefer to avoid writing one if possible :)

mattklein123 commented:
@jhspaybar not that I know of. @jmarantz do you know of anything?

jmarantz commented:
Sorry, haven't dealt with certs much.

mattklein123 commented:
@jmarantz sorry the question is about code to do URL encoding.

jmarantz commented:
Sorry, yes...I have some stuff to point to.  It's a multi-dimensional space...code to escape arbitrary binary characters into a query-parameter value is here:

https://github.com/apache/incubator-pagespeed-mod/blob/f39340ef8c30dd1a2b09e7154d30c83b887b20c5/pagespeed/kernel/http/google_url.cc#L629

That's not a bad way to go if you have something that's mostly alphanumeric, but with a few characters that need to be escaped to live in a query-param value.  The advantage is that browsers can decode for JS, if that's a goal.  And the alphanumeric input characters will stay as is and human-readable.

However if it's really binary data (eg certs?), it will be pretty bloated, and may grow too large for a URL in some contexts.  Base64 in theory should do better, but depending on which base64 encoder the result may still need to be escaped.  In the past I've used something that I've called "web64" which is the same concept but can be directly included in URLs or query parameters.  Code is here:

https://github.com/apache/incubator-pagespeed-mod/blob/f39340ef8c30dd1a2b09e7154d30c83b887b20c5/third_party/base64/base64.cc#L188

web64 is base64 but drawn from this charset:

             "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
             "abcdefghijklmnopqrstuvwxyz"
             "0123456789-_";

staying away from + and /.

Hope this helps.

PiotrSikora commented:
@jmarantz `base64(DER)` is what we have now, but the suggestion is to do `urlencode(PEM)` to provide more portable format (but PEM contains new lines, so we need to urlencode it).

btw: your "web64" is called "base64url" :)

jmarantz commented:
I see...PEMs are mostly base64-encoded already.  A special-case URL-encoder for PEM should be possible that can recover the PEM perfectly without really escaping anything.  It might look like
  ?privateKey=base64urlencodedbytes&cert=base64urlencodedbytes
where the re-encoder just replaced / and = with - and _, and removed newlines in each section.  The decoder would re-insert newlines every N chars, inserted the section markers. transform - and _ back to / and -.

Is that what you want?  That would be very minimal but not totally standard.  A pure query-param encoder like the one in google_url.cc:629 is probably what you want, and will mostly not explode the content of the PEM.

I'm still not sure where this is going to show up...in an inter-proxy request?

PiotrSikora commented:
@jmarantz No, that wouldn't be portable. It's going to be sent to and consumed by backends.

jmarantz commented:
Left some comments in slack, as I'm not sure I understand all the requirements.

What medium will the PEM be transmitted in (request header, URL, protobuf)
What software will decode it (envoy?  the backend job envoy runs?)

However if you want a standard URL encoder, google_url.cc:629 has one (can be decode by browsers).  It all depends on who is decoding the URLs.

jhspaybar commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +230,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  RELEASE_ASSERT(bio_buf->length != 0);
+  std::string pem = std::string(bio_buf->data, bio_buf->length);
+  // URL encoding shortcut
+  pem = std::regex_replace(pem, std::regex("\n"), "%0A");
```
This and the following lines look potentially expensive to me.  Would I be better off iterating the string once and just switching on every character before pushing it onto the end of a second string?  I was worried a bit that char iteration could cause issues with different charsets, but I assume it'll be safe here since Im already dealing with a pem formatted string?

jhspaybar commented:
I've addressed all the comments, and may want to tweak the approach to URL encoding the pem string, but will wait to hear back before doing so.

I also still have some concerns about injection attacks, purposeful or accidental through the Subject.  If it contains a `;` I'm worried parsing could go badly.  @mattklein123 had pinged @myidpt @lizan for info on what Istio does and I hadn't seen this addressed yet, so wanted to bring it up one more time.

myidpt commented:
Sorry for the late reply.

For the ";" issue (appearing inside subject), we could do special coding, or use base64 encoding as you mentioned, whichever is more performant.

For preventing injection attacks, I agree with Matt. I think we don't offer protection for that. In Istio with mTLS, each workloads needs to trust all the previous hops to set the header correctly. But in an unsecured network, anything can happen so there's no secure way to do it.

For preventing injection attacks, (I'm just thinking aloud) one possible solution is to attach the private-key signature to the xfcc header at each hop. The workload needs to verify every previous hop's signature to make sure that there's no tempering of the xfcc header through the hops. This is obviously very costly and the workload may not be accessible to the pub keys of previous hops.

jhspaybar commented:
It looks like the failures are related to a RateLimitIntegrationTest ?

```[ RUN      ] IpVersionsClientType/RatelimitIntegrationTest.Error/0
[       OK ] IpVersionsClientType/RatelimitIntegrationTest.Error/0 (201 ms)
[ RUN      ] IpVersionsClientType/RatelimitIntegrationTest.Timeout/0
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:101] Caught Segmentation fault, suspect faulting address 0x0
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:85] Backtrace obj</build/tmp/_bazel_bazel/e1c4f70ea3ec3e18fe575296c4ce4334/execroot/ci/bazel-out/k8-dbg/bin/test/coverage/coverage_tests.runfiles/ci/test/coverage/coverage_tests> thr<0> (use tools/stack_decode.py):
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #0 0x1df1ee9
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #1 0x3906354
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #2 0x38fe37c
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #3 0x38da85d
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #4 0x38dbd3f
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #5 0x38dc9f8
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #6 0x38ec53d
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #7 0x3907bf5
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #8 0x38ffb79
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #9 0x38e99d1
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #10 0x2f99918
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #11 0x2f99d55
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #12 0x2f97886
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:93] thr<0> obj</lib/x86_64-linux-gnu/libc.so.6>
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #13 0x7f5e4f1f082f
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:93] thr<0> obj</build/tmp/_bazel_bazel/e1c4f70ea3ec3e18fe575296c4ce4334/execroot/ci/bazel-out/k8-dbg/bin/test/coverage/coverage_tests.runfiles/ci/test/coverage/coverage_tests>
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:95] thr<0> #14 0x42c2f8
[2018-01-24 20:53:03.361][9985][critical][backtrace] bazel-out/k8-dbg/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:97] end backtrace thread 0
Segmentation fault
================================================================================
Target //test/coverage:coverage_tests up-to-date:```

and 

```@envoy//test/tools/router_check/test:router_tool_test                    PASSED in 14.2s
@envoy//test/integration:ratelimit_integration_test                      FAILED in 101.4s
  /build/tmp/_bazel_bazel/400406edc57d332f0b9b805d2b8e33a1/execroot/envoy/bazel-out/k8-dbg/testlogs/external/envoy/test/integration/ratelimit_integration_test/test.log

Executed 190 out of 190 tests: 189 tests pass and 1 fails locally.```  This seems unrelated to me.

dnoe commented:
You should be able to get the symbols resolved in that stack trace using the steps documented here: https://github.com/envoyproxy/envoy/blob/master/bazel/README.md#stack-trace-symbol-resolution

jhspaybar commented:
It looks related to the flakiness in the rate limit test fixed earlier today.  I'll try rebasing on master and kicking off the CI stuff again.

mattklein123 commented:
@dnoe @PiotrSikora do you mind taking a pass over this since @ggreenway is out this week?

PiotrSikora commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +228,32 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  RELEASE_ASSERT(buf != nullptr);
+  RELEASE_ASSERT(PEM_write_bio_X509(buf.get(), cert.get()) == 1);
+  BUF_MEM* bio_buf = nullptr;
+  BIO_get_mem_ptr(buf.get(), &bio_buf);
```
This should be replaced with `BIO_mem_contents()`.

PiotrSikora commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +228,32 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  // URL encoding shortcut
+  std::vector<std::pair<const absl::string_view, std::string>> replacements;
+  replacements.push_back({"\n", "%0A"});
+  replacements.push_back({"=", "%3D"});
```
Nit: move this to the end to sort replacements "alphabetically".

PiotrSikora commented on source/common/http/conn_manager_impl.h:
```diff
@@ -160,7 +160,7 @@ enum class ForwardClientCertType {
  * information to the next hop.
  */
-enum class ClientCertDetailsType { Subject, SAN };
+enum class ClientCertDetailsType { Subject, SAN, Cert };
```
Nit: this should probably be first item, not the last.

jhspaybar replied:
I assume there's no worry about changing the number values then, or should I be explicit and do { Cert=2, Subject=0, SAN=1 }?

mattklein123 replied:
Changing the values is fine.

PiotrSikora commented on test/common/ssl/ssl_socket_test.cc:
```diff
@@ -545,9 +550,49 @@ TEST_P(SslSocketTest, GetSubjectsWithBothCerts) {
+           "0AIwQYMBaAFDt4pFFPFoSTHEgoegytK5ZByn15MA0GCSqGSIb3DQEBCwUAA4GBAI2q%"
+           "0AKKsieXM9jr8Zthls1W83YIcquaO9XLnKFRZwfQk4yU3t7erQwAroq9wXm6T6NS23%0AoHhHNYIF91JP%"
+           "2BA9jcY2rJCTKibTUk21mVxrmr9qxKhAPJyhWoaAnEoVBgU9R9%2Bos%"
+           "0AARHpgiMhyCDvnWCdHY5Y64oVyiWdL9aHv5s82GrV%0A-----END%20CERTIFICATE-----%0A",
```
Nit: could you retain full lines here (i.e. `%0A` at the end of line and don't merge lines) to improve readability?

jhspaybar commented:
@PiotrSikora thanks for the review comments, I think I've addressed them now.

PiotrSikora commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +228,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  bssl::UniquePtr<BIO> buf(BIO_new(BIO_s_mem()));
+  RELEASE_ASSERT(buf != nullptr);
+  RELEASE_ASSERT(PEM_write_bio_X509(buf.get(), cert.get()) == 1);
+  const uint8_t* output = nullptr;
```
Nit: I don't think you need to initialize this.

PiotrSikora commented on include/envoy/ssl/connection.h:
```diff
@@ -15,39 +15,45 @@ class Connection {
   virtual std::string uriSanPeerCertificate() PURE;
+
+  /**
+   * @return std::string the PEM encoded representation of the peer certificate. Returns "" if
```
URL-encoded PEM-encoded...

PiotrSikora commented on include/envoy/ssl/connection.h:
```diff
@@ -15,39 +15,45 @@ class Connection {
+   * @return std::string the PEM encoded representation of the peer certificate. Returns "" if
+   *         there is no peer certificate or encoding fails.
+   **/
+  virtual std::string peerCertificate() const PURE;
```
I have mixed feelings about exposing URL-encoded PEM-encoded peer certificate under this function name (vs raw peer certificate - if we ever need one).

I'm not strongly against it, but wanted to see if anyone else thinks this should be named differently.

PiotrSikora commented on include/envoy/ssl/connection.h:
```diff
@@ -50,10 +50,10 @@ class Connection {
+   *         "" if there is no peer certificate or encoding fails.
    **/
-  virtual std::string peerCertificate() const PURE;
+  virtual std::string urlEncodedPemEncodedCertificate() const PURE;
```
Nit: you've lost "peer" to indicate that it's remote and not local certificate.

mattklein123 commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +228,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
   return Hex::encode(computed_hash);
 }
 
+std::string SslSocket::urlEncodedPemEncodedPeerCertificate() const {
```
As an aside, I think that for perf reasons we could cache the result of this inside this class. For kept alive connections, it's very likely this will end up getting called many times. I would just keep that in mind for the future if this shows up on perf traces. Can you add a TODO?

jhspaybar replied:
Sure, I was planning to ask *how* to do this after I landed this first PR.  I agree that many of these values should be cached, but subjects/other things also weren't caching and I figured I might hit all of them in the same second PR.

mattklein123 commented on source/common/ssl/ssl_socket.cc:
```diff
@@ -227,6 +228,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
+  RELEASE_ASSERT(BIO_mem_contents(buf.get(), &output, &length) == 1);
+  std::string pem = std::string(reinterpret_cast<const char*>(output), length);
+  // URL encoding shortcut
+  std::vector<std::pair<const absl::string_view, std::string>> replacements;
```
I would use initializer list init here and make the variable const. Something like:
```
const std::vector<std::pair<const absl::string_view, std::string>> replacements{{"\n", "%0A"}, ...};
```

jhspaybar commented:
@mattklein123 I addressed the nits.  If there is a "right way" to keep a small cache in this class for envoy, I'd be happy to follow up with a PR that adds it.  Please give me some pointers if you can :)  Once I finish adding the cache in a second PR, I'll still need to go back and document the new proto option as well.

mattklein123 commented:
Let's just go ahead and merge this and do a followup for caching. What I would do is something quite simple. Basically:
1) Inside the `SslSocket` class, add private member variables for each `std::string` you would like to cache. (You might actually want to use `Optional<std::string>` so you know if empty string is the cached value if that applies).
2) In each relevant function, check if the string has been created, if not, do the creation and cache it. Then just return the value.
3) Note that this has the benefit of the fact that you should be able to change the functions to return `const std::string&` which after the first call will then return by reference and avoid the copy.

jhspaybar commented:
Oh!  I guess there are no threading or sharing concerns?  I was expecting to need to use something like a guava loading cache at a static level on the class so it were shared more widely. Is there only a single instance of these ssl socket classes per connection?

mattklein123 commented:
> Oh! I guess there are no threading or sharing concerns? I was expecting to need to use something like a guava loading cache at a static level on the class so it were shared more widely. Is there only a single instance of these ssl socket classes per connection?

Yes. I think I would just do a cache per connection. Pretty simple.

htuch commented:
It looks like this PR merge has caused CI failure:

```
external/envoy/test/common/ssl/ssl_socket_test.cc:1909:3: error: no matching function for call to 'testUtil'
  testUtil(revoked_client_ctx_json, server_ctx_json, "", "", "", "", "", "ssl.fail_verify_error",
  ^~~~~~~~
external/envoy/test/common/ssl/ssl_socket_test.cc:42:6: note: candidate function not viable: requires 11 arguments, but 10 were provided
void testUtil(const std::string& client_ctx_json, const std::string& server_ctx_json,
     ^
external/envoy/test/common/ssl/ssl_socket_test.cc:1919:3: error: no matching function for call to 'testUtil'
  testUtil(successful_client_ctx_json, server_ctx_json, "", "", "", "", "", "ssl.handshake", true,
  ^~~~~~~~
external/envoy/test/common/ssl/ssl_socket_test.cc:42:6: note: candidate function not viable: requires 11 arguments, but 10 were provided
void testUtil(const std::string& client_ctx_json, const std::string& server_ctx_json,
     ^
2 errors generated.
INFO: Elapsed time: 728.969s, Critical Path: 88.30s
```

This seems a legit bug; @jhspaybar can you take a look at this? Thanks.

jhspaybar commented:
Sure. It looks like some tests were added that didn’t cause a merge conflict so my branch stayed green since I hadn’t rebased off master in a couple days. I’ll get you a PR in about an hour.

alyssawilk commented:
Yeah, I saw the same thing.  We'd best unbreak the build as fast as possible though, as it makes reviews harder and merges tricky.
https://github.com/envoyproxy/envoy/pull/2490 should fix

jhspaybar commented:
Thanks @alyssawilk looks like you beat me to it!

mattklein123 commented:
@jhspaybar I forgot, but when you do your doc PRs, can you also add a note to https://github.com/envoyproxy/envoy/blob/master/RAW_RELEASE_NOTES.md. Thanks!
